Refactor gin middleware (#224)
airbrake/gobrake: examples/gin/README.md

# Gin integration

This is an example of a basic Gin app with Airbrake middleware that reports performance data (route stats).

## How to run Example API

Insert your project ID and project key in the `main.go` file. You can find these values on the settings page for your project.

> Initialise mod file

Reviewer: Add a code block explaining how to do this.

```
go mod init
go mod tidy
```

Run go application

The example application provides three GET endpoints:

1. `/date` - gets the system date from the server
2. `/locations` - gets the supported locations for use with the `/weather` endpoint
3. `/weather/{locationName}` - gets the weather for a location; valid values for `locationName` can be found using the `/locations` endpoint

Use the cURL commands below to interact with the endpoints. The endpoints require an `api-key` HTTP header.

```bash

> curl "http://127.0.0.1:3000/date" -H 'api-key: d4b371692d361869183d92d84caa5edb8835cf7d'

Reviewer: ```suggestion
curl "http://localhost:3000/date" -H 'api-key: d4b371692d361869183d92d84caa5edb8835cf7d'
```

This is an example of basic Gin app with Airbrake middleware that reports route stats.

> ## How to run Exampe API

Reviewer 2: ```suggestion
## How to run Example API
```

Copy the example in your testing folder.

> Insert youy project ID and project key in main.go file

Reviewer 2: ```suggestion
Insert your project ID and project key in main.go file
```

run:

```bash

> go mod init {directory-name}

Reviewer 2: I think we could provide a test name so that users don't have to think hard about it.

go mod tidy
go run main.go
```

Valid endpoints for testing are:

`curl "http://127.0.0.1:3000/date" -H 'api-key: d4b371692d361869183d92d84caa5edb8835cf7d'`

`curl "http://127.0.0.1:3000/locations" -H 'api-key: d4b371692d361869183d92d84caa5edb8835cf7d'`

`curl "http://127.0.0.1:3000/weather/{austin/pune/santabarbara}" -H 'api-key: d4b371692d361869183d92d84caa5edb8835cf7d'`

> Once you call the API end-points, check your project dashboard for performance

Reviewer 2: ```suggestion
Once you call the API endpoints, check your project's performance dashboard
```

> This is an example of basic Gin app with Airbrake middleware that reports route stats.

Reviewer: ```suggestion
This is an example of a basic Gin app with Airbrake middleware that reports performance data (route stats).
```

## How to run Example API

> Copy the example in your testing(eg. ginbrake) folder.

Reviewer 2: ```suggestion
Copy the example in your testing (e.g. ginbrake) folder.
```

> Copy the example in your testing(eg. ginbrake) folder.

Reviewer: What is a "testing (e.g. ginbrake) folder"? Does a user really need to copy it? or can they just run it from the example directory?

> Insert your project ID and project key in main.go file

Reviewer: ```suggestion
Insert your project ID and project key in the `main.go` file. You can find these values on the settings page for your project.
```

> run:

Reviewer: ```suggestion
Run the following commands to build and run this example application.
```

```bash

> mkdir ginbrake

Reviewer: Why do we need to create a new `ginbrake` directory? Can the user just:

1. Update `main.go` to include project ID and key
2. Initialize go mod
3. Run main.go

```bash
# Add project ID and key to main.go
go mod init
go mod tidy
go run main.go
```

Author: This is to give developers the choice not to clone the whole repository and just the example application. Otherwise they would have to clone the repository or find where the gobrake package is installed and traverse to gin sample application.

Reviewer 2: How can you clone just the example application?

cd ginbrake
go mod init ginbrake
go mod tidy
go run main.go
```

> Valid endpoints for testing are:

Reviewer: ```suggestion
The sample application provides three GET endpoints:

1. `/date` - gets the system date from the server
2. `/locations` - gets the supported locations for use with the `/weather` endpoint
3. `/weather/{locationName}` - gets the weather for a location; valid values for `locationName` can be found using the `/locations` endpoint

Use the cURL commands below to interact with the endpoints. The endpoints require an `api-key` HTTP header.
```

> `curl "http://127.0.0.1:3000/date" -H 'api-key: d4b371692d361869183d92d84caa5edb8835cf7d'`

Reviewer: ```suggestion
`curl "http://localhost:3000/date" -H 'api-key: d4b371692d361869183d92d84caa5edb8835cf7d'`
```

Reviewer: Also, put all the curl statements in a single code block rather than separate code blocks:

```
curl "..."
curl "..."
curl "..."
```

Instead of:

`curl ...`

`curl ...`

`curl ...`

`curl "http://127.0.0.1:3000/locations" -H 'api-key: d4b371692d361869183d92d84caa5edb8835cf7d'`

`curl "http://127.0.0.1:3000/weather/{austin/pune/santabarbara}" -H 'api-key: d4b371692d361869183d92d84caa5edb8835cf7d'`

> 

Reviewer: Add another curl statement, something like this:

To demonstrate the Airbrake error handling, make a request to the `/weather` endpoint with an unsupported location.

```
curl "http://localhost:3000/weather/boston" -H 'api-key: d4b371692d361869183d92d84caa5edb8835cf7d'
```

The response should be a `404 Not Found` HTTP response. The example application also considers this an error and notifies the Airbrake service with the error information.

> Once you call the API endpoints, check your project's performance dashboard

Reviewer: ```suggestion
Once you call the API endpoints, view the Airbrake errors and performance dashboards for your project.
```
